// @nx/react-native/src/generators/application/lib/add-e2e.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addE2e = addE2e;
const add_e2e_1 = require("@nx/react/src/generators/application/lib/add-e2e");
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../../../utils/versions");
async function addE2e(host, options) {
    switch (options.e2eTestRunner) {
        case 'cypress':
            return (0, add_e2e_1.addE2e)(host, {
                ...options,
                e2eTestRunner: 'cypress',
                style: 'none',
                styledModule: null,
                hasStyles: false,
                unitTestRunner: 'none',
                names: (0, devkit_1.names)(options.name),
            });
        case 'playwright':
            return (0, add_e2e_1.addE2e)(host, {
                ...options,
                e2eTestRunner: 'playwright',
                style: 'none',
                styledModule: null,
                hasStyles: false,
                unitTestRunner: 'none',
                names: (0, devkit_1.names)(options.name),
            });
        case 'detox':
            const { detoxApplicationGenerator } = (0, devkit_1.ensurePackage)('@nx/detox', versions_1.nxVersion);
            return detoxApplicationGenerator(host, {
                ...options,
                e2eName: options.e2eProjectName,
                e2eDirectory: options.e2eProjectRoot,
                appProject: options.projectName,
                appDisplayName: options.displayName,
                appName: options.name,
                framework: 'react-native',
                setParserOptionsProject: options.setParserOptionsProject,
                skipFormat: true,
            });
        case 'none':
        default:
            return () => { };
    }
}
