// @nx/expo/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.babelRuntimeVersion = exports.jestExpoVersion = exports.testingLibraryReactNativeVersion = exports.reactNativeSvgVersion = exports.reactNativeSvgTransformerVersion = exports.metroVersion = exports.expoMetroRuntimeVersion = exports.expoMetroConfigVersion = exports.reactNativeWebVersion = exports.reactNativeVersion = exports.typesReactVersion = exports.reactDomVersion = exports.reactVersion = exports.babelPresetExpoVersion = exports.expoCliVersion = exports.expoSystemUiVersion = exports.expoStatusBarVersion = exports.expoSplashScreenVersion = exports.expoVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.expoVersion = '~53.0.10';
exports.expoSplashScreenVersion = '~0.30.9';
exports.expoStatusBarVersion = '~2.2.3';
exports.expoSystemUiVersion = '~5.0.8';
exports.expoCliVersion = '~0.24.14'; // @expo/cli
exports.babelPresetExpoVersion = '~13.2.0';
exports.reactVersion = '19.0.0';
exports.reactDomVersion = '19.0.0';
exports.typesReactVersion = '~19.0.10';
exports.reactNativeVersion = '0.79.3';
exports.reactNativeWebVersion = '~0.20.0';
exports.expoMetroConfigVersion = '~0.20.14';
exports.expoMetroRuntimeVersion = '~5.0.4';
exports.metroVersion = '~0.82.4';
exports.reactNativeSvgTransformerVersion = '~1.5.1';
exports.reactNativeSvgVersion = '~15.11.2';
exports.testingLibraryReactNativeVersion = '~13.2.0';
exports.jestExpoVersion = '~53.0.7';
exports.babelRuntimeVersion = '~7.27.6';
