// @nx/react-native/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.babelRuntimeVersion = exports.reactNativeSvgWebVersion = exports.reactNativeSvgVersion = exports.reactNativeSvgTransformerVersion = exports.reactTestRendererVersion = exports.testingLibraryReactNativeVersion = exports.typesReactDomVersion = exports.typesReactVersion = exports.reactDomVersion = exports.reactVersion = exports.metroVersion = exports.reactNativeWebVersion = exports.reactNativeMetroConfigVersion = exports.reactNativeBabelPresetVersion = exports.reactNativeCommunityCliPlatformIosVersion = exports.reactNativeCommunityCliPlatformAndroidVersion = exports.reactNativeCommunityCliVersion = exports.typesNodeVersion = exports.reactNativeVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.reactNativeVersion = '~0.79.3';
exports.typesNodeVersion = '20.19.9';
exports.reactNativeCommunityCliVersion = '~18.0.0';
exports.reactNativeCommunityCliPlatformAndroidVersion = '~18.0.0';
exports.reactNativeCommunityCliPlatformIosVersion = '~18.0.0';
exports.reactNativeBabelPresetVersion = '~0.79.3';
exports.reactNativeMetroConfigVersion = '~0.79.3';
exports.reactNativeWebVersion = '~0.20.0';
exports.metroVersion = '~0.82.4';
exports.reactVersion = '19.0.0';
exports.reactDomVersion = '19.0.0';
exports.typesReactVersion = '~19.0.10';
exports.typesReactDomVersion = '~19.0.6';
exports.testingLibraryReactNativeVersion = '~13.2.0';
exports.reactTestRendererVersion = '~19.0.0';
exports.reactNativeSvgTransformerVersion = '~1.5.1';
exports.reactNativeSvgVersion = '~15.11.2';
exports.reactNativeSvgWebVersion = '~1.0.9';
exports.babelRuntimeVersion = '~7.27.6';
